src/inmanta/ast/__init__.py

```diff
@@ -465,6 +467,29 @@ def get_namespace(self) -> "Namespace":
     def get_location(self) -> Location:
         return self.location
 
+    def warn_shadowed_variables(self, subblock: Optional["BasicBlock"] = None) -> None:
```

Comment:
Why is this method on namespace if it is only concerned with blocks?

Reply:
This is something I went back and forth with. It's more immediately concerned with blocks, but on the other hand, when checking implementation blocks, it makes sense to check them against the namespace they live in instead of fetching that namespace's scope's block and checking against that.